onnxscript/type_annotation.py

```diff
@@ -28,31 +28,91 @@
+        # Present in Python 3.9+
+        if get_origin(typeinfo) is typing.Annotated:
+            return get_args(typeinfo)[0]
+    return typeinfo
+
+
+def is_primitive_attr_type(typeinfo) -> bool:
+    return typeinfo in _PYTYPE_TO_ATTRTYPE_MAP
+
+
+def pytype_to_attrtype(pytype) -> typing.Optional[onnx.AttributeProto.AttributeType]:
+    pytype = remove_annotation(pytype)
     if pytype in _PYTYPE_TO_ATTRTYPE_MAP:
         return _PYTYPE_TO_ATTRTYPE_MAP[pytype]
-    if get_origin(pytype) in _LIST_CONSTRUCTORS:
-        args = get_args(pytype)
-        elt_type = args[0]
+    type_constructor = get_origin(pytype)
+    # Remove Optional wrapper if present, which is represented as an Union[..., None]
+    if type_constructor is typing.Union:
+        # Filter out None, since typing.Optional[X] evaluates to Union[X, None]
+        args = [x for x in get_args(pytype) if x is not type(None)]  # noqa: E721
+        if len(args) == 1:
+            return pytype_to_attrtype(args[0])
+    if type_constructor in _LIST_CONSTRUCTORS:
+        elt_type = get_args(pytype)[0]
         if elt_type in _LISTTYPE_TO_ATTRTYPE_MAP:
             return _LISTTYPE_TO_ATTRTYPE_MAP[elt_type]
     return None
 
 
-def is_attr_type(pytype: type):
-    return pytype_to_attrtype(pytype) is not None
-
-
-def is_value_type(typeinfo):
+def is_tensor_type(typeinfo):
     if isinstance(typeinfo, TensorType):
         return True
     if inspect.isclass(typeinfo) and issubclass(typeinfo, TensorType):
         return True
     return False
 
 
+def is_value_type(typeinfo) -> bool:
+    """Returns True if typeinfo represents a value type, False if it is an attribute type.
+    Raises ValueError if typeinfo is not a supported type annotation.
+    """
+    typeinfo = remove_annotation(typeinfo)
+    if is_tensor_type(typeinfo):
+        return True
+    if is_primitive_attr_type(typeinfo):
+        return False
+    type_constructor = get_origin(typeinfo)
+    # Handle List-like type-constructor
+    # Eg. List[INT32] is a value type, while List[int] is an attribute type
+    if type_constructor in _LIST_CONSTRUCTORS:
+        elt_type = get_args(typeinfo)[0]
+        return is_value_type(elt_type)
+    # Handle Union and Optional type-constructors
+    if type_constructor is typing.Union:
+        # Filter out None, since typing.Optional[X] evaluates to Union[X, None]
+        args = [x for x in get_args(typeinfo) if x is not type(None)]  # noqa: E721
+        args_value_check = [is_value_type(x) for x in args]
+        if all(args_value_check):
+            # Handles cases like Optional[INT32] as well as Union[FLOAT16, FLOAT, DOUBLE]
+            return True
+        elif (len(args) == 1) and args_value_check[0] is False:
+            # Handle the case of optional attribute: eg. Optional[int]
+            # Note that we do not allow Union[int, float] for attributes.
+            return False
+        else:
+            raise ValueError(f"Unsupported type annotation {typeinfo}")
```

Comment:
```suggestion
            raise ValueError(f"Unsupported type annotation '{typeinfo}'")
```

Quote user inputs for clarity? Here and elsewhere